File: sem4/jpp/l4/cpp/main.cpp
#include <iostream>
#include <mutex>
#include <chrono>
#include <thread>
#include <random>

const int ilosc_filozofow = 5;
const int min_dobranoc = 50;
const int max_dobranoc = 100;
const int ilosc_posilkow = 5;


class Filozof {
  private:
    int id;
    std::mutex* left;
    std::mutex* right;
    int ilosc_posilkow;
    
  public:
    Filozof(int id, std::mutex &left, std::mutex &right, int ilosc_posilkow){
      this->id = id;
      this->left = &left;
      this->right = &right;
      this->ilosc_posilkow = ilosc_posilkow;
    }


    void operator()() {
      this->start();
    }

    void mysl() {
      std::cout << "Filozof " << this->id << " mysli \n";
      this->dobranoc();
      std::cout << "Filozof " << this->id << " skonczyl myslec \n";
    }

    void jedz() {
      std::cout << "Filozof " << this->id << " je \n";
      this->dobranoc();
      std::cout << "Filozof " << this->id << " skonczyl jesc \n";
    } 

    void podnies_widelec(std::mutex &widelec) {
      dobranoc();
      widelec.lock();
    }

    void odloz_widelec(std::mutex &widelec) {
      dobranoc();
      widelec.unlock();
    }

    void dobranoc() {
      std::random_device rd;
      std::mt19937 rng(rd());
      std::uniform_int_distribution<int> uni(min_dobranoc, max_dobranoc);
      std::this_thread::sleep_for(std::chrono::milliseconds((uni(rng) + max_dobranoc)));
    }

    void start() {
      for(int i = 0; i < this->ilosc_posilkow; i++){
        this->mysl();
        this->podnies_widelec(*this->left);
        this->podnies_widelec(*this->right);
        this->jedz();
        this->odloz_widelec(*this->left);
        this->odloz_widelec(*this->right);
      }
    }
};


int main() {
  std::vector<std::thread> filozofowie(ilosc_filozofow);
  std::vector<std::mutex> widelce(ilosc_filozofow);

  for(int i = 0; i < ilosc_filozofow; i++) {
    auto& left = widelce[i];
    auto& right = widelce[(i + 1) % ilosc_filozofow];
    Filozof filozof = (i % 2 == 0) ? Filozof(i, right, left, ilosc_posilkow) : Filozof(i, left, right, ilosc_posilkow);
    filozofowie[i] = std::thread(filozof);
  }
  for (auto &filozof : filozofowie) {
    filozof.join();
  }

  return 0;
}

File: sem4/jpp/l4/go/main.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

func main() {
	var widelce = [ilosc_filozofow]Widelec{}
	var filozofowie = [ilosc_filozofow]*Filozof{}
	var waitGroup = sync.WaitGroup{}

	for i := 0; i < ilosc_filozofow; i++ {
		var lewa = &widelce[i]
		var prawa = &widelce[(i+1)%ilosc_filozofow]

		var filozof *Filozof
		if i%2 == 0 {
			filozof = nowyFilozof(i, prawa, lewa, ilosc_posilkow, &waitGroup)
		} else {
			filozof = nowyFilozof(i, lewa, prawa, ilosc_posilkow, &waitGroup)
		}
		filozofowie[i] = filozof
	}

	for _, f := range filozofowie {
		waitGroup.Add(1)
		go f.start()
	}
	waitGroup.Wait()
}

const (
	ilosc_filozofow int = 5

	min_dobranoc   int           = 50
	max_dobranoc   int           = 100
	jeden_dzien    time.Duration = time.Millisecond
	ilosc_posilkow int           = 5
)

var rng = rand.New(rand.NewSource(time.Now().UnixNano()))

type Filozof struct {
	id             int
	left           *Widelec
	right          *Widelec
	ilosc_posilkow int
	waitGroup      *sync.WaitGroup
}

func nowyFilozof(id int, left *Widelec, right *Widelec, ilosc_posilkow int, waitGroup *sync.WaitGroup) *Filozof {
	return &Filozof{id, left, right, ilosc_posilkow, waitGroup}
}

func (f *Filozof) mysl() {
	fmt.Printf("Filozof %d mysli \n", f.id)
	f.dobranoc()
	fmt.Printf("Filozof %d skonczyl myslec \n", f.id)
}

func (f *Filozof) jedz() {
	fmt.Printf("Filozof %d je \n", f.id)
	f.dobranoc()
	fmt.Printf("Filozof %d skonczyl jesc \n", f.id)
}

func (f *Filozof) podnies_widelec(widelec *Widelec) {
	f.dobranoc()
	widelec.Lock()
}

func (f *Filozof) odloz_widelec(widelec *Widelec) {
	f.dobranoc()
	widelec.Unlock()
}

func (f *Filozof) dobranoc() {
	time.Sleep(time.Duration(rng.Intn(min_dobranoc)+max_dobranoc) * jeden_dzien)
}

func (f *Filozof) start() {
	defer f.waitGroup.Done()
	for i := 0; i < f.ilosc_posilkow; i++ {
		f.mysl()
		f.podnies_widelec(f.left)
		f.podnies_widelec(f.right)
		f.jedz()
		f.odloz_widelec(f.left)
		f.odloz_widelec(f.right)
	}
}

type Widelec struct {
	sync.Mutex
}
